OCPBUGS-49728: Adapt MCC to use LayeredNodeState and remove LayeredPoolState

djoshy opened the pull request:
**- What I did**
- Refactored all node related functions to work with the `LayeredNodeState` object
- Removed all functions and tests related to layered pools as we no longer annotate images to `MachineConfigPools`
- Reworked some of the status update checks to use the `LayeredNodeState` struct
- Updated a few unit tests to guard for https://issues.redhat.com/browse/OCPBUGS-43552

I have a couple of open questions, for which I'll leave comments below!

**- How to verify it**
- Existing units/e2es should pass.
- Please stress test this PR for OCL rollout/status behaviors as quite a bit of that has been refactored and I may have missed something.

djoshy commented on pkg/controller/node/status.go:
```diff
@@ -359,95 +359,14 @@ func isNodeManaged(node *corev1.Node) bool {
-		if mosb != nil && mosc != nil {
-			mosbState := ctrlcommon.NewMachineOSBuildState(mosb)
-			// It seems like pool image annotations are no longer being used, so node specific checks were required here
-			if layered && mosbState.IsBuildSuccess() && mosb.Spec.MachineConfig.Name == pool.Spec.Configuration.Name && isNodeDoneAt(node, pool, layered) && lns.IsCurrentImageEqualToBuild(mosc) {
```
I've updated this check to remove `MachineOSBuild` successful condition comparison, since I think this can be solely determined by the node's annotation against the MOSB and MOSC objects while layered. Happy to add it back if I'm missing something here!

djoshy commented on pkg/controller/common/layered_node_state.go:
```diff
@@ -38,125 +42,119 @@ func (l *LayeredNodeState) IsDoneAt(mcp *mcfgv1.MachineConfigPool, layered bool)
-	return val == "" || !ok
+	// If the MOSC does not have an image, but the node has an older image annotation, the image is still likely
+	// being built.
+	return false
```
Wanted to double check this thought: This function will now only be called when the node is layered. As a result, if the image annotation is missing for whatever reason, we should always return false. In the old implementation, this was used in non layered mode too, so it return true when the annotation was missing.

djoshy commented on pkg/controller/node/node_controller.go:
```diff
@@ -1184,30 +1188,10 @@ func (ctrl *Controller) updateCandidateNode(mosc *mcfgv1.MachineOSConfig, mosb *
-			// when using this new system
-			// we know the mosc+mosb can trigger one another and cause a build, but if the node controller
-			// can't set this anno, and subsequently cannot trigger the daemon to update, we need to rework.
 			lns.SetDesiredStateFromMachineOSConfig(mosc, mosb)
```
I removed these checks because:

1. The `getAllCandidateMachines` function already returns a list of viable candidate nodes, which does this by comparing the node's annotation against the MCP/OCL objects as required.
2. This list is then passed to `updateCandidateMachines` which truncates the list based on the pool's capacity.
3. Finally, the list is passed to the above `updateCandidateNode` which was essentially re-doing the checks `getAllCandidateMachines` performed. 
 
This seemed excessive, so I removed those checks from this function to simplify things and make it easier to read.

djoshy commented:
/tide refresh
/jira refresh

cheesesashimi reviewed (commented):
Overall, some really nice changes here! I have a couple of ideas to make it even better, most of which are just renaming things and using some helper functions built into `kubernetes/apimachinery`.

I also included some thought exercises intended to challenge the status quo. I tried to be as clear as possible that I'm not asking for it to be implemented right now and that it is purely a thought exercise only. If there's any question about that, please ask :smile:

Again, nicely done with all of this!

djoshy commented on pkg/controller/common/layered_node_state.go:
```diff
@@ -38,125 +42,119 @@ func (l *LayeredNodeState) IsDoneAt(mcp *mcfgv1.MachineConfigPool, layered bool)
-func (l *LayeredNodeState) SetDesiredStateFromPool(layered bool, mcp *mcfgv1.MachineConfigPool) {
+func (l *LayeredNodeState) SetDesiredStateFromPool(mcp *mcfgv1.MachineConfigPool) {
 	node := l.Node()
 	if node.Annotations == nil {
```
Neat, will do!

djoshy commented on pkg/controller/node/node_controller.go:
```diff
@@ -1218,6 +1202,12 @@ func (ctrl *Controller) updateCandidateNode(mosc *mcfgv1.MachineOSConfig, mosb *
 			return err
 		}
 
+		// Don't make a patch call if no update is needed.
```
Good to know :smile:

djoshy commented on pkg/controller/common/layered_node_state.go:
```diff
@@ -38,125 +42,119 @@ func (l *LayeredNodeState) IsDoneAt(mcp *mcfgv1.MachineConfigPool, layered bool)
-	return val == "" || !ok
+// Checks if the current image annotation is present on the node.
+func (l *LayeredNodeState) IsCurrentImageAnnotationPresentOnNode() bool {
+	_, desiredOK := l.node.Annotations[daemonconsts.CurrentImageAnnotationKey]
```
Will do!

djoshy commented on pkg/controller/common/layered_node_state.go:
```diff
@@ -25,10 +25,14 @@ func NewLayeredNodeState(n *corev1.Node) *LayeredNodeState {
+	if layered {
+		return l.IsNodeDone() && l.IsDesiredMachineConfigEqualToPool(mcp) && l.IsDesiredEqualToBuild(mosc, mosb)
+	}
+	return l.IsNodeDone() && l.IsDesiredMachineConfigEqualToPool(mcp) && !l.IsCurrentImageAnnotationPresentOnNode()
```
Hmm, I think both of these might actually be covered by the image annotation check in the [`isNodeDone()` call](https://github.com/djoshy/machine-config-operator/blob/fe8353e4ea7e72dfd69105069b870a37a87478ec/pkg/controller/common/layered_node_state.go#L197)

djoshy commented on pkg/controller/common/layered_node_state.go:
```diff
@@ -38,125 +42,119 @@ func (l *LayeredNodeState) IsDoneAt(mcp *mcfgv1.MachineConfigPool, layered bool)
 // the opposite of getReadyNodes().
 //
-// This augments this check by determining if the desired iamge annotation is
+// This augments this check by determining if the desired image annotation is
```
Good callout! Will update.

djoshy commented on pkg/controller/common/layered_node_state.go:
```diff
@@ -38,125 +42,119 @@ func (l *LayeredNodeState) IsDoneAt(mcp *mcfgv1.MachineConfigPool, layered bool)
-// specified by the pool.
-func (l *LayeredNodeState) IsDesiredEqualToPool(mcp *mcfgv1.MachineConfigPool, layered bool) bool {
-	return l.isDesiredMachineConfigEqualToPool(mcp) && l.isDesiredImageEqualToPool(mcp, layered)
+func (l *LayeredNodeState) IsUnavailable() bool {
```
Will do!

djoshy commented on pkg/controller/common/layered_node_state_test.go:
```diff
@@ -1,358 +0,0 @@
-package common
```
Can do :smile:

djoshy commented:
Updated with:
- Brought back units for `LayeredNodeState`, and adapted the tests to work with our new functions
- Some function renames that @cheesesashimi suggested
- Added metav1 helpers that for annotation checks

djoshy commented:
/retest-required

djoshy commented:
/test unit

djoshy commented:
/test unit

cheesesashimi commented:
Nicely done with this! I apologize that my final review took so long. In light of recent discussions around OCL, an idea came to mind that I just wanted to capture someplace:

A given node can be in a number of different states. For example, if a MachineConfig annotation changes but the MCD hasn't yet reacted to it, that's one state. Once the MCD starts working, it transitions from that state to another state. Where I'm going with this is that I think we should eventually enumerate all of these states and determine which states a node can transition to as part of a state machine. This would also help us better understand how MachineConfigs and images get rolled out across a given pool.

Just to be clear, I'm not asking you to figure out any of that as part of this PR since it is a broader discussion that I think we need to have.

djoshy commented:
/cherry-pick release-4.19
